<a href="https://colab.research.google.com/github/anayamgreen/A01-data-wrangling/blob/main/A01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling

### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing?

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** Anaya Green

**Date:** 09/02/2026

In [77]:
# Your Phase 1 solution to the problem goes here.

##-------------- Question 1 of Phase 1-------------------
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#read in the csv using pd.read_csv
df = pd.read_csv('./airbnb_hw.csv')
var = df['Price']
# look at the numbers of rows and colimns using .shape
print("var.shape:")
print(var.shape, '\n')


#  look at each column types using .dtypes
print("var.dtypes:")
print(var.dtypes, '\n')

# Look at the first few rows of the data using .head()
var.head()
print(var.unique(), '\n')
print(var.value_counts(), '\n')


#Diagnosisng the dirty code

var = var.str.replace(",","") # removing the commas

var =pd.to_numeric(var, errors='coerce') # making the Price columnn numeric instead of an object

#print(var.dtypes)# checking the data types

df['var_nan'] =var.isnull().sum() # looking at if there is any null or missing values
print("Total Missing Values:", df['var_nan'])


print('Before coercion:\n', df["Price"].describe(), '\n')
print('After coercion:\n', var.describe(), '\n')
print('Total missings:\n', sum(df['var_nan']), '\n')




var.shape:
(30478,) 

var.dtypes:
object 

['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1,000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1,200' '158' '650' '234' '310' '695' '400' '166'
 '119' '6

**How many missing values do you end up with?**
There are no missing values after the ddata type change



In [76]:
##-------Question 2 Phase 1-----------
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# Read in csv using pd.read_csv
df = pd.read_csv('./mn_police_use_of_force.csv')


# look at the number of rows and columns using .shape
print("df.shape:")
print(df.shape, '\n')


# look at the data value types using .dtypes
print("df.dtypes:")
print(df.dtypes, '\n')

# look at the dataframe first few rows with .head()
df.head()


# assigning the subject_injury column to the var variable
var = 'subject_injury'


# use .unique() to look at the specific data without its replicates
print("Replication:",df[var].unique(), '\n')


#value_counts() counts the amount of the specific values and returns the
# frequency that those values are reported
print("Count the amount of specific values", df[var].value_counts(), '\n')


# Checking for the amount of missing values using .isnull().sum()
print("The amount of missing values for the subject_injury:", df[var].isnull().sum())


# Eliminating the missing values using .dropna() and assigning it to a new variable
df_cleaned = df.dropna(subset=[var])


# use .unique() to look at the specific data without its replicates and checking to
#see if missing values have been removed
print("New data:", df_cleaned[var].unique(), '\n')
print("The cleaned data:", df_cleaned)


# cross tabulation between subject_injury and force_type
var2 = df['force_type']

print(var2.isnull().sum)# checking for any missing values

cross = pd.crosstab(df_cleaned[var], var2, dropna = False)

# Checking for any missing values
cross_na = cross.isnull()
print("Check for any missing values:",cross_na)

df.shape:
(12925, 13) 

df.dtypes:
response_datetime     object
problem               object
is_911_call           object
primary_offense       object
subject_injury        object
force_type            object
force_type_action     object
race                  object
sex                   object
age                  float64
type_resistance       object
precinct               int64
neighborhood          object
dtype: object 

Replication: [nan 'No' 'Yes'] 

Count the amount of specific values subject_injury
Yes    1631
No     1446
Name: count, dtype: int64 

The amount of missing values for the subject_injury: 9848
New data: ['No' 'Yes'] 

The cleaned data:         response_datetime                   problem is_911_call  \
9     2016/01/01 02:38:23                    Fight           No   
10    2016/01/01 02:38:23                    Fight           No   
11    2016/01/01 02:39:47        Suspicious Person           No   
12    2016/01/01 02:39:47        Suspicious Person           No   
1

# Phase 1 Question 2: Written Response

**What proportion of the values are missing? Is this a concern?**

There are 9,848 missing values and 12,925 data values in the subject_injury column; therefore, approximately 76.2% of the data is missing. This is a concern because this represents more than half of the data recorded for the subject_injury column.

**Cross-tabulate your cleaned subject_injury variable with the force_type variable. Are there any patterns regarding when the data are missing?**

There are no missing values in the cross tabulation


## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** [To be provided]
- **AI tool and model used:** [To be provided]

### *Prompts used

1. [Paste prompt]
2. [Paste follow-up prompt]

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. ...
2. ...

### *Evaluation of AI suggestions


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]